File: labs/calculator/calculator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// adds/subtracts/multiplies all values that are in the *values array.
// nValues is the number of values you're reading from the array
// operator will indicate if it's an addition (1), subtraction (2) or
// multiplication (3)
long calc(int operator, int nValues, int *values) {
    long res = 0;
    if (operator == 1){ // add
        for (int i = 0; i < nValues; i++){
            res += values[i];
        }
    } else if (operator == 2){ // sub
        res = values[0];
        for (int i = 1; i < nValues; i++){
            res -= values[i];
        }
    } else if (operator == 3){ // mult
        res = 1;
        for (int i = 0; i < nValues; i++){
            res *= values[i];
        }
    }
    return res;
}

int main(int argc, char **argv) {
    char *op = argv[1];
    int *values;
    int i = 0;
    if (argc <= 3){
        printf("ERROR: Not enough arguments\n");
        return -1;
    }
    for (int j = 2; j < argc; j++){
        int val = atoi(argv[j]);
        if (val == 0 && strcmp(&argv[j], "0") != 0) { // chech if val is not an int
            printf("ERROR: %s is not an int\n", argv[j]);
            return -1;
        }
        values[i] = val;
        i++;
    }
    int size = argc - 2;
    long res;
    if (strcmp(op, "add") == 0){ // operator is add
        res = calc(1, size, values);
        for (int i = 0; i < size - 1; i++){
            printf("%d + ", values[i]);
        }
        printf("%d = %d\n", values[size - 1], res);
    } else if (strcmp(op, "sub") == 0){ // operator is sub
        res = calc(2, size, values);
        for (int i = 0; i < size - 1; i++){
            printf("%d - ", values[i]);
        }
        printf("%d = %d\n", values[size - 1], res);
    } else if (strcmp(op, "mult") == 0){ // operator is mult
        res = calc(3, size, values);
        for (int i = 0; i < size - 1; i++){
            printf("%d * ", values[i]);
        }
        printf("%d = %d\n", values[size - 1], res);
    } else { // invalid operator
        printf("ERROR: %s is an Invalid Operator\n", op);
        return -1;
    }
    return 0;
}

File: labs/go-calculator/calculator.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
)

// adds/subtracts/multiplies all values that are in the *values array.
// nValues is the number of values you're reading from the array
// operator will indicate if it's an addition (1), subtraction (2) or
// multiplication (3)
func calc(operator int, values []int) int {
	res := 0
	if operator == 1 { // add
		for i := 0; i < len(values); i++ {
			res += values[i]
		}
	} else if operator == 2 { // sub
		res = values[0]
		for i := 1; i < len(values); i++ {
			res -= values[i]
		}
	} else if operator == 3 { // mult
		res = values[0]
		for i := 1; i < len(values); i++ {
			res *= values[i]
		}
	}
	return res
}

// parse the value to int and catch errors
func parse(x string) int {
	val, err := strconv.Atoi(x)
	if err != nil { // parse error
		log.Fatalf("ERROR: %s is not an int", x)
	}
	return val
}

func main() {
	args := len(os.Args[1:])
	if args < 3 {
		log.Fatal("ERROR: Not enough arguments")
	}
	op := os.Args[1]
	size := len(os.Args[2:])
	values := make([]int, size)
	i := 0
	for j := 2; i < size; j++ {
		values[i] = parse(os.Args[j])
		i++
	}
	if op == "add" { // operator is add
		res := calc(1, values)
		for i := 0; i < size-1; i++ {
			fmt.Printf("%d + ", values[i])
		}
		fmt.Printf("%d = %d\n", values[size-1], res)
	} else if op == "sub" { // operator is sub
		res := calc(2, values)
		for i := 0; i < size-1; i++ {
			fmt.Printf("%d - ", values[i])
		}
		fmt.Printf("%d = %d\n", values[size-1], res)
	} else if op == "mult" { // operator is mult
		res := calc(3, values)
		for i := 0; i < size-1; i++ {
			fmt.Printf("%d * ", values[i])
		}
		fmt.Printf("%d = %d\n", values[size-1], res)
	} else { // invalid operator
		log.Fatalf("ERROR: %s is an Invalid Operator\n", op)
	}
}
